src/safeposix/syscalls/fs_calls.rs

```diff
@@ -910,6 +909,421 @@ impl Cage {
+                },
+                Inode::Dir(ref mut dir_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        dir_inode.mode = (dir_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+                Inode::Pipe(ref mut general_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        general_inode.mode = (general_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+                Inode::Socket(ref mut general_inode) => {
+                    if mode & (S_IRWXA|(S_FILETYPEFLAGS as u32)) == mode {
+                        general_inode.mode = (general_inode.mode &!S_IRWXA) | mode
+                    } else {
+                        //there doesn't seem to be a good syscall error errno for this
+                        return syscall_error(Errno::EACCES, "chmod", "provided file mode is not valid");
+                    }
+                    //persist metadata
+                    return 0;
+                },
+            }
+        } else {
+            return syscall_error(Errno::ENOENT, "chmod", "the provided path does not exist");
+        }
+    }
+
+    //------------------------------------FLOCK SYSCALL------------------------------------
+
+    pub fn flock_syscall(&self, fd: i32, operation: i32) -> i32 {
+        //TO DO: Implement advisory lock
+        let fdtable = self.filedescriptortable.read().unwrap();
+        match fdtable.get(&fd) {
+            Some(filedesc) => {
+                //if anything besides allowed flags, then fail
+                if 0 == operation & !(LOCK_SH|LOCK_EX|LOCK_NB|LOCK_UN) {
+                    return syscall_error(Errno::EINVAL, "flock", "operation is invalid");
+                }
+                if 0 == operation & LOCK_SH {
+                    //TO DO: raise unimplemented error
+                    return syscall_error(Errno::ENOSYS, "flock", "this operation has not been implemented yet");
+                }
+                //check whether the lock is blocking or not
+                if operation & LOCK_EX == 0 && operation & LOCK_NB == 0 {
+                    //if lock.acquire(False) then return 0, else syscall error
+                    return 0;
+                } else if 0 == operation & LOCK_EX {
+                    //fd lock acquire(true)
```

Comment:
still needs this it looks like

Reply:
right -- I am not sure how to implement the flag and don't want to screw it up.

Reply:
Our idea is that you add a field to every filedesc type (i.e. File, Pipe, etc.) which contains a interface::RustLock<()>, and for ex lock it for writing, for sh lock it for reading.
For nb use try_read/try_write instead of read/write locking, and you'll need to deal with the result. of the try, returning an errno if the try result is an Err (maybe using an if_let).
If you have any further questions, feel free to ask!